### PR TITLE
Passphrase, Password: Implemented secure random

### LINKED ISSUE #3509: Random Passphrase: Need to seed random number generator in `handle` function
It looks like it gives the same output each time: https://v.usetapes.com/0CPjlCD0Lo

Is this something that can be fixed by `srand()` like in https://github.com/duckduckgo/zeroclickinfo-goodies/blob/3cc434822c56407779d8cc07b0fa6be7cc6b96e9/lib/DDG/Goodie/Password.pm?

CC @moollaza 

------ 

IA Page: https://duck.co/ia/view/passphrase

### PR DESCRIPTION
**Pull request by ianiD**
## Description of new Instant Answer, or changes

The old Passphrase implementation used the builtin random generator and didn't even seed it on every request. Therefore, you'd get the same passphrase on multiple requests. This PR implements a random function based on [Crypt::URandom](https://metacpan.org/pod/Crypt::URandom).

While fixing the Passphrase goodie I also noticed the Password one [suffers from a similar problem](https://github.com/duckduckgo/zeroclickinfo-goodies/issues/3509#issuecomment-310729022) and plugged the same random function in there, though I'd love a recomendation on how I should only define the function once as it currently is copy-pasted from one file to the other (please don't get mad 😞).

## Related Issues and Discussions

Fixes #3509 

## People to notify

@kirkins 

## Testing & Review
To be completed by Community Leader (or DDG Staff) when reviewing Pull Request

**Pull Request**
- [ ] Title follows correct format (Specifies Instant Answer + Purpose)
- [ ] Description contains a valid Instant Answer Page Link (e.g. https://duck.co/ia/view/my_ia)

**Instant Answer Page** (for new Instant Answers)
- [ ] Instant Answer page is correctly filled out and contains:
    - The **Title** is appropriately named and formatted
    - The **IA topics** are present and appropriate
    - The **Description** is clear and coherent
    - **Source Name** exists if applicable
    - All **Example Queries** trigger on [Beta](https://beta.duckduckgo.com/)

**Code**
- [ ] Adheres to the [DuckDuckGo Style Guide](https://docs.duckduckhack.com/resources/code-style-guide.html)
- [ ] Behaviour is appropriately tested. If improvement, tests are adequately extended.
- [ ] There is no unnecessary files in place (such as editor config files)
- [ ] There is no API keys / secrets present
- [ ] Tests are passing (run `$ duckpan test <goodie_id>`)
    - Tester should report any failures

**Ready to merge?**

- [ ] Has this IA been deployed to and tested on [beta.duckduckgo.com](https://beta.duckduckgo.com/)?
- [ ] For larger commits, has this been approved be more than one community member?
- [ ] The number of reviews is appropriate for this type of PR
- [ ] The commit message is clear, coherent and fitting

**Pull Request Review Guidelines**: https://docs.duckduckhack.com/programming-mission/pr-review.html

---

Instant Answer Page: https://duck.co/ia/view/passphrase
https://duck.co/ia/view/password

### CODE REVIEW

**Comment by daxtheduck** [comment]
Metadata placeholder post
            
*If you haven't already, please edit your first comment to include a link to the IA page at the bottom.*
            
---
            
**How to find your IA page**
- New Instant Answer
  -- follow the steps [here](https://duck.co/ia/new_ia) to create your IA page.
              
- Existing Instant Answer
  -- find your IA page from this [list](https://duck.co/ia)

**Comment by pjhampton** [comment]
Thanks ianiD!
@duckduckgo/duckduckhack-contributors anybody want to review this?
@duckduckgo/community-leaders is this dependency installed on the production machines?

**Comment by kirkins** [comment]
LGTM, I could only test with `duckpan query` & `duckpan test` not sure why but `duckpan server` isn't working for me right now. I assume you tested with the server option to check that it doesn't return the same answer when viewed from the browser?

The old version never had the issue when viewed from `duckpan query` either, only in the browser.

Note to anyone who wants to test you have to run `cpan`, hit enter a few times, and then` install Crypt::URandom`.

**Comment by kirkins** [comment]
@pjhampton posted before I saw your message. I'm not sure if it's installed on production, itt's at least not installed by default on Codio.

**Comment by pjhampton** [comment]
In that case it will have to be vetted by staff. I will flag it here 👍  Thanks @kirkins

File: lib/DDG/Goodie/Passphrase.pm [hunk 1]
```diff
@@ -57,4 +61,19 @@ handle query_lc => sub {
     };
 };
 
+sub saferandom {
+    my ($range) = @_;
+
+    my $copy = $range;
+    my $rand = 0;
+    while($copy) {
+        my $byte = ord(urandom(1));
```
Comment (mintsoft) [inline lib/DDG/Goodie/Passphrase.pm]: The integer that `urandom` takes is the number of bytes to return, any particular reason not to do `urandom ($rangeinbytes)` or whatever? `ceil(log($range)/log(2))` should give you the number of bytes needed

Comment (ianiD) [inline lib/DDG/Goodie/Passphrase.pm]: Then you'd have to use `unpack` which is a bit messy. I tried it like that initially and couldn't get it working properly ¯\_(ツ)_/¯

Comment (mintsoft) [inline lib/DDG/Goodie/Passphrase.pm]: Well true, however given that we know that range isn't ever that large, we can assume that the answer will fit within a long, so you might as well replace all of this with:  `return unpack("L", urandom(8)) % $range;`

Comment (ianiD) [inline lib/DDG/Goodie/Passphrase.pm]: Yeah, I guess you're right. Should I make another commit or something?

Comment (mintsoft) [inline lib/DDG/Goodie/Passphrase.pm]: Yeah if you commit to the same branch, github will automatically add it onto the pr

Comment (ianiD) [inline lib/DDG/Goodie/Passphrase.pm]: Pushed those changes @mintsoft

File: lib/DDG/Goodie/Passphrase.pm [hunk 2]
```diff
@@ -39,7 +42,7 @@ handle query_lc => sub {
     my @chosen_words;
     while (scalar @chosen_words < $word_count) {
         # Pick random words from the slurped array until we have enough
-        push @chosen_words, $word_list[int(rand $list_size)];
+        push @chosen_words, $word_list[unpack("L", urandom(4)) % $list_size];
```
Comment (mintsoft) [inline lib/DDG/Goodie/Passphrase.pm]: I'd probably wrap this up into a sub like you have on the other one, it's a little cleaner that way

Comment (ianiD) [inline lib/DDG/Goodie/Passphrase.pm]: Done!

**Comment by mintsoft** [comment]
@moollaza this LGTM, can you look into getting the dependency onto prod?

**Comment by moollaza** [comment]
Deploying to Beta. Will need to add this dep internally, but first I want to ensure this works across multiple processes, as we have in production.

**Comment by cngarrison** [comment]
Good morning,

On 27 Jun 2017, at 8:36, Zaahir Moolla wrote:

> Deploying to Beta. Will need to add this dep internally, but first I 
> want to ensure this works across multiple processes, as we have in 
> production.

I’m just returning from holiday after finishing last contract, so 
I’m not familiar with this whole discussion. But from quick skim it 
seems to be the same issue we were discussing last year regarding how 
perl’s native `srand` works, and how it effects getting same results 
from multiple processes.

I think there are still some open GH issues on this topic with fairly 
in-depth discussion of different solutions, along with how the different 
solutions affect production.

I’ll read the history of this PR, and if relevant I’ll find the the 
other issue # and include it here for further discussion/review.

Thanks,
Charlie

-- 

     Charlie Garrison                   <charlie@garrison.com.au>
     Garrison Computer Services      <http://www.garrison.com.au>
     PO Box 380
     Tumbarumba NSW 2653  Australia

[Conundrum](http://www.ietf.org/rfc/rfc1855.txt)

**Comment by mintsoft** [comment]
@moollaza looking at the module source, on linuxes it'll only be basically `cat /dev/urandom` (https://metacpan.org/source/DDICK/Crypt-URandom-0.36/lib/Crypt/URandom.pm#L101) so it should be totally cross-process safe anyway

**Comment by moollaza** [comment]
@ianiD Thanks! This LGTM :+1:
